### PR TITLE
Remove DoFHandlerType from dofs/dof_tools

### PR DESCRIPTION
**Pull request by peterrum**
references #10333

### CODE REVIEW

**Comment by peterrum** [comment]
Note that I had to introduce 3 deprecated functions (the reason is that the template arguments `dim` and `spacedim` cannot be deduced from the iterator):

https://github.com/dealii/dealii/blob/8db216fd783f0c7230164d9b2eabe6d50a0cb990/include/deal.II/dofs/dof_tools.h#L1121-L1140

https://github.com/dealii/dealii/blob/8db216fd783f0c7230164d9b2eabe6d50a0cb990/include/deal.II/dofs/dof_tools.h#L1934-L1941

https://github.com/dealii/dealii/blob/8db216fd783f0c7230164d9b2eabe6d50a0cb990/include/deal.II/dofs/dof_tools.h#L2277-L2284

File: include/deal.II/dofs/dof_tools.h [hunk 1]
```diff
@@ -1107,7 +1128,16 @@ namespace DoFTools
...
+    const number periodicity_factor = 1.)
+  {
+    make_periodicity_constraints<DoFHandlerType::dimension,
+                                 DoFHandlerType::space_dimension>(
+      periodic_faces,
+      constraints,
+      component_mask,
+      first_vector_components,
+      periodicity_factor);
+  }
```
Comment (kronbichler) [inline include/deal.II/dofs/dof_tools.h:1140]: Can you move the implementation to the end of this file like we do in the other functions of this file?

File: include/deal.II/dofs/dof_tools.h [hunk 2]
```diff
@@ -1868,13 +1905,8 @@ namespace DoFTools
...
    * system that results retains the block structure of the global system.
    *
-   * @tparam DoFHandlerType A type that is either DoFHandler or
-   * hp::DoFHandler. In C++, the compiler can not determine the type of
-   * <code>DoFHandlerType</code> from the function call. You need to specify
-   * it as an explicit template argument following the function name.
-   *
    * @param patch A collection of cells within an object of type
-   * DoFHandlerType
+   * DoFHandler<dim, spacedim>
```
Comment (kronbichler) [inline include/deal.II/dofs/dof_tools.h:1909]:
```suggestion
   * DoFHandler<dim, spacedim>::active_cell_iterator
```

File: include/deal.II/dofs/dof_tools.h [hunk 3]
```diff
@@ -1887,10 +1919,26 @@ namespace DoFTools
...
   get_dofs_on_patch(
-    const std::vector<typename DoFHandlerType::active_cell_iterator> &patch);
+    const std::vector<typename DoFHandler<dim, spacedim>::active_cell_iterator>
+      &patch);
+
+  /**
+   * The same as above.
+   *
+   * @deprecated Use the function that takes dim and spacedim as template
+   *   argument in favor of the following function.
```
Comment (kronbichler) [inline include/deal.II/dofs/dof_tools.h:1932]:
```suggestion
   *   argument.
```
(technically, that other function comes before).

File: include/deal.II/dofs/dof_tools.h [hunk 4]
```diff
@@ -1887,10 +1919,26 @@ namespace DoFTools
...
+   */
+  template <typename DoFHandlerType>
+  DEAL_II_DEPRECATED std::vector<types::global_dof_index>
+                     get_dofs_on_patch(
+                       const std::vector<typename DoFHandlerType::active_cell_iterator> &patch)
+  {
+    return get_dofs_on_patch<DoFHandlerType::dimension,
+                             DoFHandlerType::space_dimension>(patch);
+  }
```
Comment (kronbichler) [inline include/deal.II/dofs/dof_tools.h:1941]: Move to the end of the file.

File: include/deal.II/dofs/dof_tools.h [hunk 5]
```diff
@@ -2219,10 +2262,26 @@ namespace DoFTools
...
+   *   argument in favor of the following function.
+   */
+  template <typename DoFHandlerType>
+  DEAL_II_DEPRECATED unsigned int
+  count_dofs_on_patch(
+    const std::vector<typename DoFHandlerType::active_cell_iterator> &patch)
+  {
+    return count_dofs_on_patch<DoFHandlerType::dimension,
+                               DoFHandlerType::space_dimension>(patch);
+  }
```
Comment (kronbichler) [inline include/deal.II/dofs/dof_tools.h:2284]: Same here, move it to the inline section.

**Comment by peterrum** [comment]
Done!

**Pull request merged**